File: peer-review/TTK4145-2025-code-peer-review-files/5728345b/Elevator_Project/util/timer/timer.go
package timer

import (
	"time"
)

type Timer struct {
	EndTime time.Time
	Active  bool
}

func Active(inTimer Timer) bool {
	return inTimer.Active
}

func NewTimer() Timer {
	return Timer{EndTime: time.Time{}, Active: false}
}

func GetWallTime() time.Time {
	return time.Now()
}

func TimerStart(inTimer *Timer, duration time.Duration) {
	inTimer.EndTime = GetWallTime().Add(duration)
	inTimer.Active = true
}

func TimerStop(inTimer *Timer) {
	inTimer.Active = false
}

func TimerTimedOut(inTimer Timer) bool {
	return inTimer.Active && time.Until(inTimer.EndTime) <= 0
}

func TimerTimeLeft(inTimer Timer) time.Duration {
	if !inTimer.Active {
		return 0
	}
	timeLeft := inTimer.EndTime.Sub(GetWallTime())
	if timeLeft < 0 {
		return 0
	}
	return timeLeft
}

File: peer-review/TTK4145-2025-code-peer-review-files/3778b696/snapshot-[REDACTED]/Single_elevator/src/timer.rs
/*
use std::time::{Duration, Instant};
use std::sync::Mutex;

lazy_static::lazy_static! {
    static ref TIMER: Mutex<Timer> = Mutex::new(Timer::new());
}

struct Timer {
    end_time: Option<Instant>,
    active: bool,
}

impl Timer {
    pub fn new() -> Self {
        Timer {
            end_time: None,
            active: false,
        }
    }

    pub fn start(&mut self, duration: f64) {
        self.end_time = Some(Instant::now() + Duration::from_secs_f64(duration));
        self.active = true;
    }

    pub fn stop(&mut self) {
        self.active = false;
    }

    pub fn timed_out(&self) -> bool {
        self.active && self.end_time.map_or(false, |end| Instant::now() > end)
    }
}

pub fn timer_start(duration: f64) {
    let mut timer = TIMER.lock().unwrap();
    timer.start(duration);
}

pub fn timer_stop() {
    let mut timer = TIMER.lock().unwrap();
    timer.stop();
}

pub fn timer_timed_out() -> bool {
    let timer = TIMER.lock().unwrap();
    timer.timed_out()
}
*/
